Editor/HierarchyFolderIcon.cs

```diff
@@ -0,0 +1,161 @@
+            var data = tinted.GetRawTextureData<Color32>();
+            for (int index = 0, len = data.Length; index < len; index++)
+            {
+                Color32 c = data[index];
+                byte a = c.a;
+
+                Color.RGBToHSV(c, out float h, out float s, out float v);
+                c = Color.HSVToRGB(0, 0, 1);
+                c.a = a;
+ 
+                data[index] = c;
+            }
+
+            var aaa = tinted.width * tinted.height * 4;
+            var offset = 0; 
+            for (int index = 0; index < tinted.mipmapCount; index++)
+            {
+                tinted.SetPixelData(data, index, offset);
+                aaa >>= 2;
+                offset += aaa;
+            }
+            tinted.Apply();
+            
+            return tinted;
+        }
+
+        private static void Initialize()
+        {
+            if (_isInitialized)
+            {
+                return;
+            }
+
+            _closedFolderTexture = (Texture2D) EditorGUIUtility.IconContent("Folder Icon").image;
+            _openFolderTexture = (Texture2D) EditorGUIUtility.IconContent("FolderEmpty Icon").image;
+ 
+            _openFolderSelectedTexture = GetTintedTexture(_openFolderTexture);
+            _closedFolderSelectedTexture = GetTintedTexture(_closedFolderTexture);
+            
+            const BindingFlags BindingAll = BindingFlags.Public 
+              | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance;
+            
+            //public static List<SceneHierarchyWindow> GetAllSceneHierarchyWindows()
+            var assembly = typeof(SceneView).Assembly;
+
+            var sceneHierarchyWindowType = assembly.GetType("UnityEditor.SceneHierarchyWindow");
+            _getAllSceneHierarchyWindowsMethod = sceneHierarchyWindowType.GetMethod("GetAllSceneHierarchyWindows", BindingAll);
+            _sceneHierarchyProperty = sceneHierarchyWindowType.GetProperty("sceneHierarchy");
+
+            var sceneHierarchyType = assembly.GetType("UnityEditor.SceneHierarchy");
+            _treeViewProperty = sceneHierarchyType.GetProperty("treeView", BindingAll);
+            
+            var treeViewControllerType = assembly.GetType("UnityEditor.IMGUI.Controls.TreeViewController");
+            _dataProperty = treeViewControllerType.GetProperty("data", BindingAll);
+
+            var iTreeViewDataSourceType = assembly.GetType("UnityEditor.IMGUI.Controls.ITreeViewDataSource");
+            _getRowsMethod = iTreeViewDataSourceType.GetMethod("GetRows");
+            _isExpandedMethod = iTreeViewDataSourceType.GetMethod("IsExpanded", new Type[] {typeof(TreeViewItem)});
+            
+            var gameObjectTreeViewItemType = assembly.GetType("UnityEditor.GameObjectTreeViewItem");
+            _selectedIconProperty = gameObjectTreeViewItemType.GetProperty("selectedIcon", BindingAll);
+            _objectPPTRProperty = gameObjectTreeViewItemType.GetProperty("objectPPTR", BindingAll);
```

Comment:
What are the guarantees of stability for these items? I don't want to spend too much time rehacking it every time there's a new upgrade if possible.

Reply:
Additionally, can these properties be cached, or is there a concern that they may change?

Reply:
Unfortunately we cant guarantee that this code wont break on future unity versions. But I've tested this code (not exactly this version, but another version of it with support for changing icon colors) on unity 2020.1.0b6 and it still worked.

The main idea is to get the current list of TreeViewItems being displayed in the scene hierarchy at the current frame. This list represent the visible items on the hierarchy at any given time. This could mean that some game object are culled if there are too many game objects on the scene. So, this list will change any time the hierarchy is scrolled, a game object is added or removed or a game object is collapsed/expanded. 

Another important detail is that there will be a new list of TreeViewItems for each active hierarchy window and this process needs to be repeated for each window.